base_exception/models/base_exception_method.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2011 Raphaël Valyi, Renato Lima, Guewen Baconnier, Sodexis
+# Copyright 2017 Akretion (http://www.akretion.com)
+# Mourad EL HADJ MIMOUNE <mourad.elhadj.mimoune@akretion.com>
+# Copyright 2020 Hibou Corp.
+# Copyright 2023 ACSONE SA/NV (http://acsone.eu)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+import logging
+
+from odoo import _, api, models
+from odoo.exceptions import UserError
+from odoo.osv import expression
+from odoo.tools.safe_eval import safe_eval
+
+_logger = logging.getLogger(__name__)
+
+
+class BaseExceptionMethod(models.AbstractModel):
+    _name = "base.exception.method"
+    _description = "Exception Rule Methods"
+
+    def _get_main_records(self):
+        """
+        Used in case we check exceptions on a record but write these
+        exceptions on a parent record. Typical example is with
+        sale.order.line. We check exceptions on some sale order lines but
+        write these exceptions on the sale order, so they are visible.
+        """
+        return self
+
+    def _rule_domain(self):
+        """Filter exception.rules.
+        By default, only the rules with the correct model
+        will be used.
+        """
+        return [("model", "=", self._name), ("active", "=", True)]
+
+    def detect_exceptions(self):
+        """List all exception_ids applied on self
+        Exception ids are also written on records
+        """
+        rules_info = (
+            self.env["exception.rule"]
+            .sudo()
+            ._get_rules_info_for_domain(self._rule_domain())
+        )
+        all_exception_ids = []
+        rules_to_remove = {}
+        rules_to_add = {}
+        for rule_info in rules_info:
+            main_records = self._get_main_records()
+            records_with_rule_in_exceptions = main_records.filtered(
+                lambda r, rule_id=rule_info.id: rule_id in r.exception_ids.ids
```

Comment:
@lmignon, I saw this in your change. Why `rule_id` as optional param and not like this?

```suggestion
                lambda r: rule_info.id in r.exception_ids.ids
```

Reply:
https://docs.python-guide.org/writing/gotchas/#late-binding-closures